ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java

```diff
@@ -225,5 +229,134 @@ static void validateDiskCapacity(long diskCapacityInBytes) {
+     * @param partitionClass the class of the partitions desired. Can be {@code null}.
+     * @return all the partitions in {@code partitionClass} (all partitions if it is {@code null})
+     */
+    List<PartitionId> getPartitions(String partitionClass) {
+      return getPartitionsInClass(partitionClass, false);
+    }
+
+    /**
+     * Returns all the partitions that are in state {@link PartitionState#READ_WRITE} AND have the highest number of
+     * replicas in the local datacenter for the given {@code partitionClass}.
+     * <p/>
+     * Also attempts to return only partitions with healthy replicas but if no such partitions are found, returns all
+     * the eligible partitions irrespective of replica health.
+     * <p/>
+     * If {@code partitionClass} is {@code null}, gets writable partitions from all partition classes.
+     * @param partitionClass the class of the partitions desired. Can be {@code null}.
+     * @return all the writable partitions in {@code partitionClass} with the highest replica count in the local
+     * datacenter (all writable partitions if it is {@code null}).
+     */
+    List<PartitionId> getWritablePartitions(String partitionClass) {
+      List<PartitionId> writablePartitions = new ArrayList<>();
+      List<PartitionId> healthyWritablePartitions = new ArrayList<>();
+      for (PartitionId partition : getPartitionsInClass(partitionClass, true)) {
+        if (partition.getPartitionState() == PartitionState.READ_WRITE) {
+          writablePartitions.add(partition);
+          if (areAllReplicasForPartitionUp((partition))) {
+            healthyWritablePartitions.add(partition);
+          }
+        }
+      }
+      return healthyWritablePartitions.isEmpty() ? writablePartitions : healthyWritablePartitions;
+    }
+
+    /**
+     * Returns the partitions belonging to the {@code partitionClass}. Returns all partitions if {@code partitionClass}
+     * is {@code null}.
+     * @param partitionClass the class of the partitions desired.
+     * @param highestReplicaCountOnly if {@code true}, returns only the partitions with the highest number of replicas
+     *                                in the local datacenter.
+     * @return the partitions belonging to the {@code partitionClass}. Returns all partitions if {@code partitionClass}
+     * is {@code null}.
+     */
+    private List<PartitionId> getPartitionsInClass(String partitionClass, boolean highestReplicaCountOnly) {
+      List<PartitionId> toReturn = new ArrayList<>();
+      if (partitionClass == null) {
+        toReturn.addAll(allPartitions);
+      } else if (partitionIdsByClassAndLocalReplicaCount.containsKey(partitionClass)) {
+        SortedMap<Integer, List<PartitionId>> partitionsByReplicaCount =
+            partitionIdsByClassAndLocalReplicaCount.get(partitionClass);
+        if (highestReplicaCountOnly) {
+          toReturn.addAll(partitionsByReplicaCount.get(partitionsByReplicaCount.lastKey()));
```

Comment:
could replace with `toReturn.addAll(partitionsByReplicaCount.lastEntry().getValue());`

Reply:
It looks like `SortedMap` doesn't have a `lastEntry()` API (I believe that is in `NavigableMap` only).

Reply:
There is no necessity to use `NavigableMap` I think